partitura/io/importdcml.py

```diff
@@ -0,0 +1,278 @@
+    part = spt.Part("P0", "Metadata", quarter_duration=qdivs)
+
+    # Add notes and grace notes
+    for n_idx, note in enumerate(note_array):
+        if grace_mask[n_idx]:
+            # verify that staff and voice are the same for the grace note and the main note
+            note_el = spt.GraceNote(
+                grace_type="grace",
+                id="n-{}".format(note["id"]),
+                step=note["step"],
+                octave=note["octave"],
+                alter=note["alter"],
+                staff=note["staff"],
+                voice=note["voice"],
+                symbolic_duration={"type": "eighth"},
+            )
+
+            if grace_mask[n_idx - 1]:
+                prev_note = note_array[n_idx - 1]
+                for note_prev in part.iter_all(spt.GraceNote, note["onset_div"], note["onset_div"] + 1):
+                    if note_prev.id == "n-{}".format(prev_note["id"]):
+                        note_el.grace_prev = note_prev
+                        note_prev.grace_next = note_el
+                        break
+        else:
+            symbolic_duration = estimate_symbolic_duration(note["duration_div"], qdivs)
+            note_el = spt.Note(
+                    id="n-{}".format(note["id"]),
+                    step=note["step"],
+                    octave=note["octave"],
+                    alter=note["alter"],
+                    staff=note["staff"],
+                    voice=note["voice"],
+                    symbolic_duration=symbolic_duration)
+        part.add(note_el, start=note["onset_div"], end=(note["onset_div"]+note["duration_div"]))
+
+    # Curate grace notes
+    grace_note_idxs = np.where(grace_mask)[0]
+    for i, grace_el in enumerate(part.iter_all(spt.GraceNote)):
+        grace_idx = grace_note_idxs[i]
+        note = note_array[grace_idx]
+        # Find the next note in the same staff and voice
+        if not grace_mask[grace_idx+1]:
+            i = 1
+            next_note = note_array[grace_idx+i]
+            while note["staff"] != next_note["staff"] or note["voice"] != next_note["voice"]:
+                i += 1
+                next_note = note_array[grace_idx+i]
+            assert note["staff"] == next_note["staff"], "Grace note and main note must be in the same staff"
+            assert note["voice"] == next_note["voice"], "Grace note and main note must be in the same voice"
+            assert note["onset_div"] == next_note[
+                "onset_div"], "Grace note and main note must have the same onset"
+            for note in part.iter_all(spt.Note, note["onset_div"], note["onset_div"]+1):
+                if note.id == "n-{}".format(next_note["id"]):
+                    grace_el.grace_next = note
+                    break
+
+    # Find time signatures
+    time_signatures_changes = data["timesig"][data["timesig"].shift(1) != data["timesig"]].index
```

Comment:
is the timesig in the main file? or in another? Just checking that everything is ok, ignore this comment if it's not relevant

Reply:
Time signatures should be in the note file. With that line I am tracking time signature changes. With the shift time_sig[0] is set to NaN so the initial time signature is there.